designs/sam_build_cmd.rst

```diff
@@ -0,0 +1,442 @@
+In some programming languages like Javascript, this process is fairly straightforward (just zip a folder), and in
+others like Python, it is much involved.
+
+Customers using SAM CLI can easily create Lambda function code (using ``sam init``) and package their artifact as a
+zip file (using ``sam package``), but they have to handle the build process themselves. Customers usually implement
+their own build scripts or adopt other's scripts from the internet. This is where many customers fall off the cliff.
+
+
+What will be changed?
+---------------------
+In this proposal, we will be providing a new command, ``sam build``, to build Lambda functions for all programming
+languages that AWS Lambda supports. The cardinality of the problem is number of programming languages (N) times the
+number of package managers (M) per language. Hence is it is nearly impossible to natively support each combination.
+Instead, ``sam build`` will support an opinionated set of package managers for all programming languages. We will
+provide an option for customers to bring their own build commands or override the default for any programming language.
+SAM CLI will still take care of the grunt work of iterating through every function in the SAM template, figuring out
+the source code location, creating temporary folders to store built artifacts, running the build command and
+move artifacts to right location.
+
+
+Success criteria for the change
+-------------------------------
+#. Support all programming languages supported by AWS Lambda
+
+   * Nodejs with NPM
+   * Java with Maven
+   * Python with PIP
+   * Golang with Go CLI
+   * Dotnetcore with DotNet CLI
+
+
+#. Each Lambda function in SAM template gets built
+
+#. Produce stable builds (best effort): If the source files did not change, built artifacts should not change.
+
+#. Built artifacts should "just work" with ``sam local`` and ``sam package`` suite of commands
+
+#. Ability to provide arbitrary build commands for each Lambda Function runtime. This will either override the built-in
+   default or add build support for languages/tools that SAM CLI does not support (ex: java+gradle).
+
+#. Opt-in to building native dependencies that can run on AWS Lambda using Docker.
+
+#. Support one dependency manifest (ex: package.json) entire app or one per each Lambda function.
+
+#. Support out-of-source builds: ie. source code of Lambda function is outside the directory containing SAM template.
+
+#. Integrate build action with ``sam local/package/deploy`` commands so the Lambda functions will be automatically
+   built as part of the command without explicitly running the build command.
+
+
+Out-of-Scope
+------------
+#. Supports adding data files ie. files that are not referenced by the package manager (ex: images, css etc)
+#. Support to exclude certain files from the built artifact (ex: using .gitignore or using regex)
```

Comment:
We will exclude certain files & folders by default like `*.pyc` that makes sense, but we won't be able to support parsing an arbitrary `.gitignore` file. So it is possible that we don't create the slimmest zip file. But this could be a fast follow feature.

Reply:
another benefit of using NPM for node, is that this is automatically handled by NPM.

Reply:
This is interesting. It also creates a problem where we don't have a consistent experience across different programming languages (ie. Python won't respect .gitignore but Nodejs builds will). May be that is acceptable?